iavl/iavl_tree.go

```diff
@@ -41,139 +135,303 @@ func NewIAVLTree(cacheSize int, db dbm.DB) *IAVLTree {
 // Note that Save() clears leftNode and rightNode.  Otherwise,
 // two copies would not be goroutine independent.
 func (t *IAVLTree) Copy() merkle.Tree {
-	if t.root == nil {
+	//fmt.Printf("Copy ")
+
+	root := t.GetRoot(t.version)
+	if t.roots == nil || root == nil {
```

Comment:
is the first condition not sufficient? how could we have roots != nil and root == nil ?